src/audio/multiband_drc/multiband_drc.c

```diff
@@ -242,46 +251,72 @@ static int multiband_drc_init(struct processing_module *mod)
 	md->private = cd;
 	cd->multiband_drc_func = NULL;
 	cd->crossover_split = NULL;
-	/* Initialize to enabled is a workaround for IPC4 kernel version 6.6 and
-	 * before where the processing is never enabled via switch control. New
-	 * kernel sends the IPC4 switch control and sets this to desired state
-	 * before prepare.
-	 */
-	multiband_drc_process_enable(&cd->process_enabled);
 
-	/* Handler for configuration data */
 	cd->model_handler = comp_data_blob_handler_new(dev);
 	if (!cd->model_handler) {
 		comp_err(dev, "multiband_drc_init(): comp_data_blob_handler_new() failed.");
 		ret = -ENOMEM;
 		goto cd_fail;
 	}
 
-	/* Get configuration data and reset DRC state */
 	ret = comp_init_data_blob(cd->model_handler, bs, cfg->data);
 	if (ret < 0) {
 		comp_err(dev, "multiband_drc_init(): comp_init_data_blob() failed.");
-		goto cd_fail;
+		goto cd_model_fail;
 	}
+
 	multiband_drc_reset_state(&cd->state);
+	/* Initialize to enabled is a workaround for IPC4 kernel version 6.6 and
+	 * before where the processing is never enabled via switch control. New
+	 * kernel sends the IPC4 switch control and sets this to desired state
+	 * before prepare.
+	 */
+	multiband_drc_process_enable(&cd->process_enabled);
 
 	return 0;
 
-cd_fail:
+cd_model_fail:
 	comp_data_blob_handler_free(cd->model_handler);
+cd_fail:
 	rfree(cd);
 	return ret;
 }
 
+/**
+ * Free resources allocated by Multiband DRC processing component.
+ *
+ * This function releases all memory and resources associated with the
+ * multiband DRC component's operation. This includes dynamically allocated
+ * filter state instances as well as freeing up the model handler.
+ *
+ * @parameters "mod" Pointer to the processing module to be freed.
+ *
+ * @return 0 indicating success.
+ */
 static int multiband_drc_free(struct processing_module *mod)
 {
 	struct multiband_drc_comp_data *cd = module_get_private_data(mod);
 
 	comp_info(mod->dev, "multiband_drc_free()");
 
-	comp_data_blob_handler_free(cd->model_handler);
+	if (cd) {
+		struct multiband_drc_state *state = &cd->state;
+
+		/* Free emphasis/deemphasis IIR filter states for all channels */
+		for (int i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
+			multiband_drc_iir_reset_state_ch(&state->emphasis[i]);
```

Comment:
why are you resetting memory that is about to be freed? There is no value in this

Reply:
Sure , Eliminated the previously included calls to reset state structures before freeing, which were indeed redundant.